### PR TITLE
Add properties property for Template class

### PR DESCRIPTION
**Pull request by Danieladu**
Fixes: #3273
## Description
For some requirements from Composer side. Describe here:
Structure LG parsed result should cover the common key/value pairs results.

For example:
```
# template
[Activity
 Text=good
 Speak=hi
]
```

The parsed Template would contains a property named `properties`, in which would contains:
```
properties: {Text: "good", Speak:"hi", $type:'Activity'}
```

## Particular cases
- Pure Expression reference would be ignore
```
[A
 key1=value1
 ${T2()}     -> this one would be ignored
]
```
- would keep the origin list info
```
[A
 list = a | b
]
```

the properties would be:
`{list: ['a', 'b'], $type: 'A'}`

### CODE REVIEW

**Comment by boydc2014** [comment]
Composer also need the type, ie, Activity, HeroCard, like that. 

So, it would be good to have a $type property in that

File: libraries/botbuilder-lg/src/template.ts [hunk 1]
```diff
@@ -38,6 +38,11 @@ export class Template {
      */
     public templateBodyParseTree: lp.BodyContext;
 
+    /**
+     * The extended properties for the object.
+     */
+    public properties?: { [propertyName: string]: any };
```
Comment (joshgummersall) [inline libraries/botbuilder-lg/src/template.ts:44]: Prefer `Record<string, unknown>` for the type.

File: libraries/botbuilder-lg/src/templatesParser.ts [hunk 2]
```diff
@@ -478,21 +477,24 @@ export class TemplatesTransformer extends AbstractParseTreeVisitor<any> implemen
...
                 DiagnosticSeverity.Warning
             );
             this.templates.diagnostics.push(diagnostic);
         } else {
             try {
-                return this.antlrParseTemplate(templateBody, startLine);
+                const templateBodyContext = this.antlrParseTemplate(template.body, startLine);
+                if (templateBodyContext) {
+                    template.templateBodyParseTree = templateBodyContext;
+                    new TemplateBodyTransformer(template).transform();
```
Comment (joshgummersall) [inline libraries/botbuilder-lg/src/templatesParser.ts:496]: Does the constructed instance need to be captured at all? It seems odd to invoke `new` without assigning it to something.

File: libraries/botbuilder-lg/src/templatesParser.ts [hunk 3]
```diff
@@ -577,3 +579,50 @@ export class TemplatesTransformer extends AbstractParseTreeVisitor<any> implemen
         );
     }
 }
+
+class TemplateBodyTransformer extends AbstractParseTreeVisitor<any> implements LGTemplateParserVisitor<any> {
```
Comment (joshgummersall) [inline libraries/botbuilder-lg/src/templatesParser.ts:583]: Can we use a better type than `any`?

File: libraries/botbuilder-lg/tests/lg.test.js [hunk 4]
```diff
@@ -857,6 +857,16 @@ describe('LG', function () {
         assert.strictEqual(evaled, 5);
     });
 
+    it('TestProperties', function () {
+        var templates = preloaded.two;
```
Comment (joshgummersall) [inline libraries/botbuilder-lg/tests/lg.test.js:861]: `const`

File: libraries/botbuilder-lg/tests/lg.test.js [hunk 5]
```diff
@@ -857,6 +857,16 @@ describe('LG', function () {
...
+        assert.strictEqual(templates.toArray()[0].properties, undefined);
+
+        var templates = preloaded.StructuredTemplate;
```
Comment (joshgummersall) [inline libraries/botbuilder-lg/tests/lg.test.js:864]: `const`

**Review by joshgummersall — CHANGES_REQUESTED** [review]

**Review by joshgummersall — APPROVED** [review]

**Pull request merged**